### PR TITLE
Poetrybranch

### PR DESCRIPTION
**Pull request by evanna00**
**PR Checklist**

- [x] This comment contains a description of changes (with reason)
- [x] Referenced issue is linked
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Documentation in `docs` is updated
- [x] `docs/changelog.md` is updated

**Description of changes**
Use the poetry to manage the dependencies.

**Technical details**
Prevent the dependencies from breaking down.
Easier manage and package the dependencies in virtual environments.
Use a lockfile, can ensure repeatable installs.
Reduce the happening of issues https://github.com/ewels/MegaQC/issues/430

### CODE REVIEW

**Comment by evanna00** [comment]
TryPull

**Comment by multimeric** [comment]
You still need to remove the `package.json` changes from this branch. Look into using `git restore`.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        "Topic :: Scientific/Engineering :: Bio-Informatics",
+        "Topic :: Scientific/Engineering :: Visualization",
+    ]
+ packages=[
+    { include = "megaqc" },
+]
+
+[tool.poetry.dependencies]
+python = "^3.10"
+        argon2-cffi="20.1"
```
Comment (multimeric) [inline pyproject.toml:41]: Fix indentation here. You might want to use the "Format Document" command in vscode.

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        outlier-utils="^0.0.3"
+        webargs="^5.5" # do not upgrade, since it breaks parsing
+        querystring-parser="^1.2"
+        scipy="^1.5"
+        flatten_json="^0.1"
+        flapison="^0.30"
+        Flask-Migrate="^2.5"
+        environs="^9.2.0"
+        MarkupSafe="<2.1.0"
+
```
Comment (multimeric) [inline pyproject.toml:78]: Remove empty spaces here. The Format Document command should also help here.

**Review by multimeric — CHANGES_REQUESTED** [review]
You will also need to delete the `setup.py` file, which is the old Python metadata file for this repo.

**Comment by multimeric** [comment]
* [ ] Remove old setuptools files (`setup.py`)
* [ ] Update the CI (GitHub Actions)
* [ ] All the other things in the checklist above

**Comment by multimeric** [comment]
* [ ] Remove old setuptools files (`setup.py`)
* [ ] Update the CI (GitHub Actions)
* [ ] All the other things in the checklist above

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -18,13 +18,9 @@ jobs:
         with:
           python-version: ${{ matrix.python-version }}
 
-      - name: Install dependencies
```
Comment (multimeric) [inline .github/workflows/build.yaml:21]: Keep this as "install dependencies", this is still true.

File: .github/workflows/build.yaml [hunk 4]
```diff
@@ -18,13 +18,9 @@ jobs:
...
+      - name: Install poetry
         run: |
-          python -m pip install --upgrade pip
-          pip install -e .[dev]
-
-      - name: Test with pytest
```
Comment (multimeric) [inline .github/workflows/build.yaml:26]: Still need to run pytest. Don't remove this block.

File: CHANGELOG.rst [hunk 5]
```diff
@@ -3,7 +3,7 @@ Changelog
 
 .. _section-1:
 
-0.3.0
```
Comment (multimeric) [inline CHANGELOG.rst:6]: You need to add a new heading, and put your changes under that, not delete the old heading.

File: CHANGELOG.rst [hunk 6]
```diff
@@ -32,6 +32,10 @@ Breaking Changes
 
 -  Dropped support for Node 8
 
+-  `[#440]`_ Set up poetry to manage the dependencies, which might keep 
+   the dependencies from breaking down, and reduce the chances of the 
+   happening of issues `[#430]`
```
Comment (multimeric) [inline CHANGELOG.rst:37]: "issues such as #430"

File: docs/source/docs/installation/installation_dev.rst [hunk 7]
```diff
@@ -8,6 +8,7 @@ You will need:
 
 -  `node`_
 -  `Python 3`_
+-  `Poetry`_
```
Comment (multimeric) [inline docs/source/docs/installation/installation_dev.rst:11]: This is current an invalid link. Look into the RST format to learn how to make this hyperlink.

File: docs/source/docs/installation/installation_dev.rst [hunk 8]
```diff
@@ -18,18 +19,15 @@ If you’re doing development work, you need access to the source code
 
    git clone https://github.com/ewels/MegaQC
 
-2. Install the Python into a virtual environment
+2. Set up the Poetry
```
Comment (multimeric) [inline docs/source/docs/installation/installation_dev.rst:22]: "Set up ~~the~~ Poetry"

Comment (multimeric) [reply docs/source/docs/installation/installation_dev.rst:22]: Poetry is a proper noun (name), so it doesn't need "the" before it.

**Review by multimeric — CHANGES_REQUESTED** [review]

File: CHANGELOG.rst [hunk 9]
```diff
@@ -3,7 +3,7 @@ Changelog
 
 .. _section-1:
 
-0.3.0
+0.3.1
```
Comment (multimeric) [inline CHANGELOG.rst:6]: No, you need to add a new heading *only* for your own changes. The other changes need to stay under the 0.3.0 heading.

File: CHANGELOG.rst [hunk 10]
```diff
@@ -32,6 +32,15 @@ Breaking Changes
 
 -  Dropped support for Node 8
 
+.. _development-1:
+
+Development
+~~~~~~~~~~~~~~~~
+
+-  `[#440]`_ Set up poetry to manage the dependencies, which might keep
```
Comment (multimeric) [inline CHANGELOG.rst:40]: Think hyperlink is wrongly formatted. Look up the RST syntax.

File: docs/source/docs/installation/installation_dev.rst (hunk 7 above)
Comment (multimeric) [reply docs/source/docs/installation/installation_dev.rst:11]: This is now the right syntax, but you need to link to the page that describes how to install poetry. Look at the other links for inspiration.

File: docs/source/docs/installation/installation_dev.rst [hunk 11]
```diff
@@ -18,18 +19,15 @@ If you’re doing development work, you need access to the source code
 
    git clone https://github.com/ewels/MegaQC
 
-2. Install the Python into a virtual environment
+2. Set up Poetry
 ------------------------------------------------
 
-You should do your development in a virtual environment. You also need
-to install MegaQC and all its dependencies there:
+You should install the poetry. You also need to install MegaQC and all its dependencies there:
```
Comment (multimeric) [inline docs/source/docs/installation/installation_dev.rst:25]: Remove "the".

**Review by multimeric — CHANGES_REQUESTED** [review]

File: CHANGELOG.rst [hunk 12]
```diff
@@ -71,3 +71,21 @@ Internal Changes
 .. _[#156]: https://github.com/ewels/MegaQC/issues/156
 .. _[#170]: https://github.com/ewels/MegaQC/issues/170
 .. _[#194]: https://github.com/ewels/MegaQC/issues/194
+
+.. _section-2:
+
+0.3.1
```
Comment (multimeric) [inline CHANGELOG.rst:77]: Newer versions go at the top.

File: docs/source/docs/installation/installation_dev.rst (hunk 7 above)
Comment (multimeric) [reply docs/source/docs/installation/installation_dev.rst:11]: Use this URL: https://python-poetry.org/docs#installation

File: docs/source/docs/installation/installation_dev.rst [hunk 13]
```diff
@@ -18,18 +19,15 @@ If you’re doing development work, you need access to the source code
 
    git clone https://github.com/ewels/MegaQC
 
-2. Install the Python into a virtual environment
+2. Set up Poetry
 ------------------------------------------------
 
-You should do your development in a virtual environment. You also need
-to install MegaQC and all its dependencies there:
+You should install Poetry. You also need to install MegaQC and all its dependencies there:
```
Comment (multimeric) [inline docs/source/docs/installation/installation_dev.rst:25]: "install MegaQC using Poetry"

**Review by multimeric — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yaml [hunk 14]
```diff
@@ -26,6 +26,9 @@ jobs:
       - name: Test with pytest
         run: pytest
 
+      - name: Install poetry
```
Comment (multimeric) [inline .github/workflows/build.yaml:29]: Remove the `pip install` block.

Comment (multimeric) [reply .github/workflows/build.yaml:29]: Need to run `pytest` inside `poetry`

File: .github/workflows/build.yaml [hunk 15]
```diff
@@ -26,6 +26,9 @@ jobs:
       - name: Test with pytest
         run: pytest
```
Comment (multimeric) [inline .github/workflows/build.yaml:27]: Look into: https://python-poetry.org/docs/basic-usage/#using-poetry-run

File: CHANGELOG.rst [hunk 16]
```diff
@@ -1,6 +1,25 @@
...
+-----
+
+.. _development-1:
+
+Development
+~~~~~~~~~~~~~~~~
+
+-  `[#440]`_ Set up poetry to manage the dependencies, which might keep 
+   the dependencies from breaking down, and reduce the chances of the 
+   happening of issues such as `[#430]`
```
Comment (multimeric) [inline CHANGELOG.rst:16]: Fix the hyperlink for 430.

**Review by multimeric — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yaml (hunk 3 above)
Comment (multimeric) [reply .github/workflows/build.yaml:21]: @evanna00

Comment (evanna00) [reply .github/workflows/build.yaml]: I modified it just now, not pretty sure if the commands I added are correct now.

**Comment by multimeric** [comment]
I have merged a pull request into master that unfortunately means you will need to change a few things:

* [ ] Update `pytest` version
* [ ] Fix the merge conflicts

Let me know if you have trouble with either of these.

File: .github/workflows/build.yaml (hunk 3 above)
Comment (multimeric) [reply .github/workflows/build.yaml]: What I meant was that you should name the section "install dependencies", but that it should use `poetry` to do so. You should remove the `pip install` commands.

**Review by multimeric — COMMENTED** [review]
* [ ] Need to recreate `poetry.lock` with the new dependencies

**Comment by multimeric** [comment]
Looks good, we just need to fix your commits having the wrong author

**Review by multimeric — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yaml [hunk 17]
```diff
@@ -18,13 +18,16 @@ jobs:
...
-      - name: Install dependencies
-        run: |
-          python -m pip install --upgrade pip
-          pip install -e .[dev]
+      - uses: snok/install-poetry@v1
+
+      - name: Install all extra features of poetry
+        run: poetry install --extras "dev"
+
+      - name: Update setuptools
```
[the remark this answers is not in the document]
Comment (multimeric) [reply .github/workflows/build.yaml]: To be clear, you do need the `poetry install`, but you don't need the `pip install`.

[the remark this answers is not in the document]
Comment (multimeric) [reply .github/workflows/build.yaml]: @evanna00

**Review by multimeric — APPROVED** [review]

**Pull request merged**